jetty-util-ajax/src/main/java/org/eclipse/jetty/util/ajax/AsyncJSON.java

```diff
@@ -0,0 +1,1205 @@
+//
+//      The Eclipse Public License is available at
+//      http://www.eclipse.org/legal/epl-v10.html
+//
+//      The Apache License v2.0 is available at
+//      http://www.opensource.org/licenses/apache2.0.php
+//
+//  You may elect to redistribute this code under either of these licenses.
+//  ========================================================================
+//
+
+package org.eclipse.jetty.util.ajax;
+
+import java.nio.ByteBuffer;
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.concurrent.ConcurrentHashMap;
+
+import org.eclipse.jetty.util.ArrayTernaryTrie;
+import org.eclipse.jetty.util.BufferUtil;
+import org.eclipse.jetty.util.Loader;
+import org.eclipse.jetty.util.Trie;
+import org.eclipse.jetty.util.TypeUtil;
+import org.eclipse.jetty.util.Utf8StringBuilder;
+
+public class AsyncJSON
+{
+    public static class Factory
+    {
+        private Trie<String> cache;
+        private Map<String, JSON.Convertor> convertors;
+        private boolean detailedParseException;
+
+        public boolean isDetailedParseException()
+        {
+            return detailedParseException;
+        }
+
+        public void setDetailedParseException(boolean detailedParseException)
+        {
+            this.detailedParseException = detailedParseException;
+        }
+
+        public void cache(String value)
+        {
+            if (cache == null)
+                cache = new ArrayTernaryTrie.Growing<>(false, 64, 64);
+            cache.put("\"" + value + "\"", value);
+        }
+
+        public String cached(ByteBuffer buffer)
+        {
+            if (cache != null)
+            {
+                String result = cache.getBest(buffer, 0, buffer.remaining());
+                if (result != null)
+                {
+                    buffer.position(buffer.position() + result.length() + 2);
+                    return result;
+                }
+            }
+            return null;
+        }
+
+        public AsyncJSON newAsyncJSON()
+        {
+            return new AsyncJSON(this);
+        }
+
+        public void putConvertor(String className, JSON.Convertor convertor)
+        {
+            if (convertors == null)
+                convertors = new ConcurrentHashMap<>();
+            convertors.put(className, convertor);
+        }
+
+        public JSON.Convertor removeConvertor(String className)
+        {
+            if (convertors != null)
+                return convertors.remove(className);
+            return null;
+        }
+    }
+
+    private static final Object UNSET = new Object();
+
+    private final Factory factory;
+    private final FrameStack stack = new FrameStack();
+    private final Utf8StringBuilder stringBuilder;
+    private final NumberBuilder numberBuilder;
+    private List<ByteBuffer> chunks;
+
+    public AsyncJSON(Factory factory)
```

Comment:
Shouldn't this constructor be package-private?

Reply:
No, it should be `public` to allow subclassing from other packages.

Reply:
Wouldn't that be `protected`?

Reply:
If it is `protected` you cannot instantiate it directly. For example, if in the future we allow configuration of the parser instance, this would not be possible:

```java
class MyFactory extends AsyncJSON.Factory {
  @Override
  public AsyncJSON newAsyncJSON() {
    AsyncJSON result = new AsyncJSON(this); // Does not compile.
    result.setFoo("bar");
    System.err.println(result);
    return result;
  }
}
```
Sure you can call `super` to get the instance, but not if you want to customize the instance creation, etc.

I feel in this case we don't gain much with `protected`, that instead becomes most of the times just an annoyance (because I can work it around anyway, just with more code).